actioncable/lib/action_cable/connection/client_socket.rb

```diff
@@ -99,6 +99,7 @@ def close(code = nil, reason = nil)
 
         @ready_state = CLOSING unless @ready_state == CLOSED
         @driver.close(reason, code)
+        begin_close(reason, code) if force && (@driver.state != :closed)
```

Comment:
I had to add a way to force the connection to close.

When the WebSocket version is Hybi, [the driver's `#close` behaviour changes such that it never emits a close event](https://github.com/faye/websocket-driver-ruby/blob/main/lib/websocket/driver/hybi.rb#L150-L154).

Because of this the connection is never closed and just lingers:
```
13:24:46 web.1  | I, [2023-12-08T13:24:46.507291 #440625]  INFO -- : [LATENCY] [device:Device-586795065] 0.008601ms                                                                                                                                                                 
13:24:49 web.1  | I, [2023-12-08T13:24:49.505572 #440625]  INFO -- : [LATENCY] [device:Device-586795065] 0.006799ms                                                                                                                                                                 
13:24:58 web.1  | I, [2023-12-08T13:24:58.499496 #440625]  INFO -- : MISSED PONG                                                                                                                                                                                                    
13:24:58 web.1  | I, [2023-12-08T13:24:58.499582 #440625]  INFO -- : Called Connection#close with reason: heartbeat_timeout and reconnect: true                                                                                                                                     
13:24:58 web.1  | I, [2023-12-08T13:24:58.499659 #440625]  INFO -- : Called WebSocket#close                                                                                                                                                                                         
13:24:58 web.1  | I, [2023-12-08T13:24:58.502117 #440625]  INFO -- : Called SocketClient#close
13:25:01 web.1  | I, [2023-12-08T13:25:01.499804 #440625]  INFO -- : MISSED PONG
13:25:01 web.1  | I, [2023-12-08T13:25:01.499849 #440625]  INFO -- : Called Connection#close with reason: heartbeat_timeout and reconnect: true
13:25:01 web.1  | I, [2023-12-08T13:25:01.499926 #440625]  INFO -- : Called WebSocket#close
13:25:01 web.1  | I, [2023-12-08T13:25:01.501111 #440625]  INFO -- : Called SocketClient#close
```

The Driver implement three versions of the WebSocket protocol: Hybi, Draft76 & Draft75
[Draft76](https://github.com/faye/websocket-driver-ruby/blob/main/lib/websocket/driver/draft76.rb#L31) & [Draft75](https://github.com/faye/websocket-driver-ruby/blob/main/lib/websocket/driver/draft75.rb#L21) immediately change the state to `:closed`. Only [Hybi](https://github.com/faye/websocket-driver-ruby/blob/main/lib/websocket/driver/hybi.rb#L150-L154) changes it to `:closing`.

Reply:
I looked into that, but calling a private method also doesn't seem right.
I'd rather invoke the callback directly, or open up a PR against the driver to add the ability to force close a connection.

Reply:
The people over at [faye-websocket-dirver gave me a very nice explanation as to what's going on here](https://github.com/faye/websocket-driver-ruby/issues/90).

The TL;DR - The Hybi driver implements the final spec of WebSockets, which states that the client *has to respond to a close message* and therefore there is no way to force a connection closed except to close the socket and trigger any callbacks yourself. Since they just implement the protocol and don't work with sockets threads timeouts or anything similar there is nothing that they can do to force a connection to close.

They also stressed that `state` shouldn't be used. Even though it's a public method, it's only intended to be used in their test suite and is intentionally undocumented. Because of that I'll remove the `state` check when `force` is `true`.

Given what we know now I think that triggering the callback manually seems to be the best option in this particular case.